src/collectors/ceph/ceph.py

```diff
@@ -69,78 +169,357 @@ def get_default_config(self):
+            else:
+                name = name_class(_PATH_SEP.join(filter(None, [prefix] + path)))
+
+                value = lookup_dict_path(stats, path)
+
+                if stat_type & _PERFCOUNTER_TIME:
+                    value = self._ceph_time_to_seconds(value)
+                    self.publish_gauge(name, value, 6)
+
+                elif stat_type & _PERFCOUNTER_U64:
+                    # create a list of values to log. we'll either log a list
+                    # of derived metrics, or the single metric we began with.
+                    if name.endswith("bytes"):
+                        values = self._get_byte_metrics(name, value)
+                    else:
+                        values = [(name, value)]
+
+                    for name, value in values:
+                        if stat_type & _PERFCOUNTER_COUNTER:
+                            self.publish_counter(name, value, 2)
+                        else:
+                            self.publish_gauge(name, value, 2)
+                else:
+                    message = "Unexpected metric stat_type: %s/%d"
+                    self.log.error(message, name, stat_type)
+
+    def _cluster_id_prefix(self, cluster_name, fsid):
+        # We'll either use the cluster name (human friendly but may not be
+        # unique) or the UUID (robust but obscure)
+        if self.config['short_names']:
+            return cluster_name
+        else:
+            return fsid
+
+    def _publish_cluster_stats(self,
+                               cluster_name,
+                               fsid,
+                               prefix,
+                               stats,
+                               counter=False):
+        """
+        Given a stats dictionary, publish under the cluster path (respecting
+        short_names and cluster_prefix)
+        """
+
         for stat_name, stat_value in flatten_dictionary(
             stats,
-            prefix=counter_prefix,
+            path=[self._cluster_id_prefix(cluster_name, fsid), prefix]
         ):
-            self.publish_gauge(stat_name, stat_value)
+            stat_name = _PATH_SEP.join(stat_name)
+            name = GlobalName(stat_name)
+            if counter:
+                self.publish_counter(name, stat_value)
+            else:
+                self.publish_gauge(name, stat_value)
+
+    def _admin_command(self, socket_path, command):
+        try:
+            json_blob, err = _popen_check_output([self.config['ceph_binary'],
+                                                  '--admin-daemon',
+                                                  socket_path] + command)
+        except CalledProcessError, e:
+            raise AdminSocketError(socket_path, command)
+
+        try:
+            return json.loads(json_blob)
+        except (ValueError, IndexError):
+            self.log.exception('Error parsing output from %s' % socket_path)
+            raise AdminSocketError(socket_path, command)
```

Comment:
@jaingaurav I would prefer not doing that. Popen can raise ValueError when called with invalid args. It would make the self.log.exception misleading